src/Interface/ComputationalRelation.agda

```diff
@@ -5,68 +5,130 @@ module Interface.ComputationalRelation where
-    ... | just (s'' , h) = mk⇔ (λ where refl → h) λ h' →
-      begin just s''        ≡˘⟨ completeness _ _ _ _ h ⟩
+    computeFail : C → S → Sig → Set
+    computeFail c s sig = isFailure $ compute c s sig
+
+    ≡-success⇔STS : compute c s sig ≡ success s' ⇔ STS c s sig s'
+    ≡-success⇔STS {c} {s} {sig} {s'} with computeProof c s sig in eq
+    ... | success (s'' , h) = mk⇔ (λ where refl → h) λ h' →
+      begin success s''     ≡˘⟨ completeness _ _ _ _ h ⟩
             compute c s sig ≡⟨ completeness _ _ _ _ h' ⟩
-            just s'         ∎
-    ... | nothing = mk⇔ (λ ()) λ h →
-      begin nothing         ≡˘⟨ map-nothing eq ⟩
+            success s'      ∎
+    ... | failure y = mk⇔ (λ ()) λ h →
+      begin failure _       ≡˘⟨ map-failure eq ⟩
             compute c s sig ≡⟨ completeness _ _ _ _ h ⟩
-            just s'         ∎
+            success s'      ∎
 
-    nothing⇒∀¬STS : compute c s sig ≡ nothing → ∀ s' → ¬ STS c s sig s'
-    nothing⇒∀¬STS comp≡nothing s' h rewrite ≡-just⇔STS .Equivalence.from h =
+    failure⇒∀¬STS : computeFail c s sig → ∀ s' → ¬ STS c s sig s'
+    failure⇒∀¬STS comp≡nothing s' h rewrite ≡-success⇔STS .Equivalence.from h =
       case comp≡nothing of λ ()
 
-    recomputeProof : ∀ {Γ s sig s'} → STS Γ s sig s' → Maybe (∃[ s'' ] STS Γ s sig s'')
+    ∀¬STS⇒failure : (∀ s' → ¬ STS c s sig s') → computeFail c s sig
+    ∀¬STS⇒failure {c = c} {s} {sig} ¬sts with computeProof c s sig
+    ... | success (x , y) = ⊥-elim (¬sts x y)
+    ... | failure y = (y , refl)
+
+    failure⇔∀¬STS : computeFail c s sig ⇔ ∀ s' → ¬ STS c s sig s'
+    failure⇔∀¬STS = mk⇔ failure⇒∀¬STS ∀¬STS⇒failure
+
+    recomputeProof : ∀ {Γ s sig s'} → STS Γ s sig s' → ComputationResult Err (∃[ s'' ] STS Γ s sig s'')
     recomputeProof _ = computeProof _ _ _
 
-module _ {STS : C → S → Sig → S → Set} (comp : Computational STS) where
+module _ {STS : C → S → Sig → S → Set} (comp : Computational STS Err) where
 
   open Computational comp
 
   ExtendedRelSTS = ExtendedRel STS
 
   ExtendedRel-compute : ExtendedRelSTS c s sig (compute c s sig)
   ExtendedRel-compute {c} {s} {sig} with compute c s sig in eq
-  ... | just s' = Equivalence.to ≡-just⇔STS eq
-  ... | nothing = λ s' h → case trans (sym $ Equivalence.from ≡-just⇔STS h) eq of λ ()
+  ... | success s' = Equivalence.to ≡-success⇔STS eq
+  ... | failure _  = λ s' h → case trans (sym $ Equivalence.from ≡-success⇔STS h) eq of λ ()
 
-  ExtendedRel-rightUnique : ExtendedRelSTS c s sig s' → ExtendedRelSTS c s sig s'' → s' ≡ s''
-  ExtendedRel-rightUnique {s' = just x}  {just x'} h h' =
-    trans (sym $ Equivalence.from ≡-just⇔STS h) (Equivalence.from ≡-just⇔STS h')
-  ExtendedRel-rightUnique {s' = just x}  {nothing} h h' = ⊥-elim $ h' x h
-  ExtendedRel-rightUnique {s' = nothing} {just x'} h h' = ⊥-elim $ h x' h'
-  ExtendedRel-rightUnique {s' = nothing} {nothing} h h' = refl
+  open ≡-Reasoning
+
+  success≡ : ∀ {A} → ComputationResult Err A → ComputationResult Err A → Set
```

Comment:
I think this should be at the top with all the other stuff about `ComputationResult` as well. Also, it's probably better to define as equality after applying `success-maybe`. Finally, I'd name this `_≈_` or some variant thereof.

Reply:
If I move it to the top, it starts complaining that it can't figure out what `Err` is supposed to be. I think it makes sense to keep it in the module so that `Err` is bound at the module level.

Reply:
I just added the implicit `Err` argument to the equivalence in `ExtendedRel-rightUnique`